sql/schemachanger: support add column on multiregion tables

fqazi opened the pull request:
Previously, the declarative schema changer would fall back
to the legacy schema changer when multiregion tables were
encountered. This was inadequate because this meant that
add column wasn't well supported in multiregion databases.
To address this patch will add support for updating zone
configs as a part of the ADD COLUMN.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/83535)
<!-- Reviewable:end -->

ajwerner commented on pkg/sql/descmetadata/metadata_updater.go:
```diff
@@ -184,3 +188,39 @@ func (mu metadataUpdater) DeleteSchedule(ctx context.Context, scheduleID int64)
+		return err
+	}
+
+	return mu.collectionFactory.Txn(ctx, mu.ie, mu.txn.DB(),
+		func(ctx context.Context, txn *kv.Txn, descriptors *descs.Collection) error {
+
+			// Recompute any spans for the subzones.
+			if len(zone.Subzones) != 0 {
+				tbl, err := descriptors.GetImmutableTableByID(ctx, txn, id, tree.ObjectLookupFlagsWithRequiredTableKind(tree.ResolveRequireTableDesc))
+				if err != nil {
+					return err
+				}
+				zone.SubzoneSpans, err = regionutils.GenerateSubzoneSpans(mu.codec, tbl, zone.Subzones)
+				if err != nil {
+					return err
+				}
+			}
+
+			bytes, err := protoutil.Marshal(zone)
+			if err != nil {
+				return err
+			}
+			_, err = mu.ie.Exec(ctx, "update-zone", mu.txn,
+				"UPSERT INTO system.zones (id, config) VALUES ($1, $2)", id, bytes)
+			return err
+		})
```
This mixing of txns is suspect. I see it being done above, but it all makes me uneasy. A simple-enough refactoring here would be to make the `metadataUpdater` not free-standing, but instead have the MetadataUpdaterFactory take a descs.Collection and kv.Txn when being constructed. It already takes a txn. 

Honestly, the `scexec.DescriptorMetadataUpdaterFactory` ought not be defined in `scexec` but rather in `scdeps`. Then it can reference the `*descs.Collection` which is already in scope. WDYT?

I think this refactoring can be pulled out to a separate PR.

fqazi replied:
Yes, I'll split this off and get a PR out

postamar commented:
Can you please break down this commit into smaller commits? I'm finding this hard to review as it is. For example I would welcome first seeing one or more "uncontroversial" commits involving refactoring/plumbing work that lay the groundwork for the "actual change" commits. Here it's all mixed up together and hard to keep up with.

fqazi reviewed (commented):
Let me take a stab and split off the MR refactor and enum stuff from the actual guts.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83535#-:-N5p98av6LVhSdBQbHct:b-px8vfc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)



<!-- Sent from Reviewable.io -->

fqazi committed:
sql: refactor regionutils into seperate pkg for declarative schema changer support

Previously, the sql/regionutils.go contained all the logic required to help setup
and create zone con figs. These changes split off that code into its own package,
so that the declarative schema changer can access these functions without
cyclic dependencies.

Release note: None

fqazi committed:
sql/schemachanger: support setting zone configs inside MetadataUpdater

Previously, there was no way for the declarative schema changer to set
zone configs without invoking into sql, which would lead to a circular
dependency. These changes add an interface for allow the declarative
schema changes to issue zone config updates via the MetadataUpdater.

Release note: None

fqazi committed:
sql/schemachanger: add enum type values as an element

Previously, our decomposition of enum types only included
an element for the type itself. This was inadequate because
for multi-region support we need to be able to iterate over
the values for an enum. This patch adds further decomposition
for enums.

Release note: None

postamar reviewed (changes requested):
There's plenty of good stuff here but I think this all requires a bit more forethought. This review is therefore only a partial review, I didn't get far into the last commit before stopping.

> Reviewed 7 of 71 files at r1, 1 of 36 files at r2, 75 of 75 files at r3, 23 of 23 files at r4, 13 of 13 files at r5, 20 of 51 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83535#-:-N5q9Ae40L5cODz8MRXB:b3ki25w)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @fqazi, and @postamar)

___
*[`-- commits` line 9 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/83535#-N5pqnhS5y1WfeoHVDbQ):*
Should this `regionutils` package be under `ccl`? I don't know, I'm not quite sure how this works exactly. This code does seem to require an enterprise license and some of its tests are already in `partitionccl`.
___
*[`pkg/sql/partition_utils.go` line 25 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/83535#-N5pqiHk5HM0ZDjiPp9i:-N5pqiHk5HM0ZDjiPp9j:b-gtkxmp) ([raw file](https://github.com/cockroachdb/cockroach/blob/c356f3725daf58b171adedd8f8d0f54f6363d690/pkg/sql/partition_utils.go#L25)):*
> ```Go
> // GenerateSubzoneSpans see regionutils.GenerateSubzoneSpans for details.
> // This wrapper enforces licensing restrictions which are enforced else,
> // where in the declarative schema changer.
> ```

Can you reword this comment please?
___
*[`pkg/sql/descmetadata/metadata_updater.go` line 227 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83535#-N5pw8z023USjuiOaQAM:-N5pw8z023USjuiOaQAN:b-yorqoe) ([raw file](https://github.com/cockroachdb/cockroach/blob/f092d6d096e8e65c0edf1ab39540fa7d06271879/pkg/sql/descmetadata/metadata_updater.go#L227)):*
> ```Go
> }
> 
> // SetZoneConfig implements scexec.DescriptorMetadataUpdater.
> ```

nit: you forgot this method definition in the `scexec.DescriptorMetadataUpdater` interface in a later commit.
___
*[`pkg/sql/descmetadata/metadata_updater.go` line 236 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83535#-N5pwiyA91r2cQZKFrub:-N5pwiyA91r2cQZKFruc:bsewf4i) ([raw file](https://github.com/cockroachdb/cockroach/blob/f092d6d096e8e65c0edf1ab39540fa7d06271879/pkg/sql/descmetadata/metadata_updater.go#L236)):*
> ```Go
> 			"DELETE FROM system.zones WHERE id = $1", id)
> 		return err
> 	}
> ```

uber nit: having  the`nil`ness of an argument determine what the function does is arguably a bit of a code smell. I'm aware this is what `writeZoneConfig` does but consider having two methods `DeleteZoneConfig` and `UpsertZoneConfig` in this interface instead. Interfaces typically deserve a bit more polish than a private function.
___
*[`pkg/sql/descmetadata/metadata_updater.go` line 254 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83535#-N5pxgbW5gWgxh3FfK6_:-N5pxgbW5gWgxh3FfK6a:bn8fgur) ([raw file](https://github.com/cockroachdb/cockroach/blob/f092d6d096e8e65c0edf1ab39540fa7d06271879/pkg/sql/descmetadata/metadata_updater.go#L254)):*
> ```Go
> 		return err
> 	}
> 	_, err = ie.Exec(ctx, "update-zone", mu.txn,
> ```

nit: `"upsert-zone"`?
___
*[`pkg/sql/descmetadata/metadata_updater.go` line 258 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83535#-N5pzai24Fua-feZiT4f:-N5pzai24Fua-feZiT4g:b-plzszl) ([raw file](https://github.com/cockroachdb/cockroach/blob/f092d6d096e8e65c0edf1ab39540fa7d06271879/pkg/sql/descmetadata/metadata_updater.go#L258)):*
> ```Go
> 	return err
> 
> }
> ```

It would be worth adopting this method in `writeZoneConfig` in `sql/set_zone_config.go` to validate its correctness.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/region_utils.go` line 63 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/83535#-N5q8sWh9UTQbSZrK0rE:-N5q8sWh9UTQbSZrK0rF:b-u0e4hg) ([raw file](https://github.com/cockroachdb/cockroach/blob/3765f1bb1802a485f8d0f562900d05d1f5d9dc19/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/region_utils.go#L63)):*
> ```Go
> 	zc = regionConfig.ApplyZoneConfigExtensionForRegionalIn(zc, partitionRegion)
> 	return zc, err
> }
> ```

I don't know what's the right thing to do but I'm pretty sure building whole zone config protos in `scbuildstmt` isn't it. Sorry if this comment isn't practically helpful in the immediate. I think this is a sign that zone config support needs to be better thought out first.
___
*[`pkg/sql/schemachanger/scdecomp/decomp.go` line 301 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/83535#-N5q7Q1T6j6-QwMoJKf9:-N5q7Q1T6j6-QwMoJKfA:blmuf45) ([raw file](https://github.com/cockroachdb/cockroach/blob/3765f1bb1802a485f8d0f562900d05d1f5d9dc19/pkg/sql/schemachanger/scdecomp/decomp.go#L301)):*
> ```Go
> 			})
> 		}
> 	}
> ```

Since these elements now show up for any table which have zone configs, how do we know that they're correctly handled when we `DROP TABLE`? In practice they end up getting removed by the GC job and that's fine, but you should have op-rules to ensure that these elements don't emit any ops in such a scenario. This is currently missing.

I think you should break down this commit further, and perhaps this PR further as well, to add zone config support in a first step (and check that you pass all the tests) and in a subsequent step worry about adding columns in multi-region tables.
___
*[`pkg/sql/schemachanger/scdeps/build_deps.go` line 406 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/83535#-N5q8AS12gTzVBzXp2p3:-N5q8AS12gTzVBzXp2p4:bpr6xwe) ([raw file](https://github.com/cockroachdb/cockroach/blob/3765f1bb1802a485f8d0f562900d05d1f5d9dc19/pkg/sql/schemachanger/scdeps/build_deps.go#L406)):*
> ```Go
> 	}
> 	return &zone
> }
> ```

Same comment as for the metadataUpdater: try to use this logic in the existing codebase, instead of copy-pasting it. Also I don't think we want business logic in `scdeps`, this should be moved elsewhere.
___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 452 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/83535#-N5q0fsLEU6zjk6ZzeFs:-N5q0fsLEU6zjk6ZzeFt:b-kga6ao) ([raw file](https://github.com/cockroachdb/cockroach/blob/3765f1bb1802a485f8d0f562900d05d1f5d9dc19/pkg/sql/schemachanger/scpb/elements.proto#L452)):*
> ```protobuf
>   string primary_region = 3 [(gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/sql/catalog/catpb.RegionName"];
>   int32 survival_goal = 4;
>   int32 placement = 5;
> ```

These two should be enums imported from `catpb` instead of `int32`. The enum definition in turn need to be moved out of `descpb`, I think.
___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 490 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/83535#-N5q7ATuDtzNYjOotQ47:-N5q7ATuDtzNYjOotQ48:b-qlclcj) ([raw file](https://github.com/cockroachdb/cockroach/blob/3765f1bb1802a485f8d0f562900d05d1f5d9dc19/pkg/sql/schemachanger/scpb/elements.proto#L490)):*
> ```protobuf
>   uint32 table_id =1 [(gogoproto.customname) = "TableID", (gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/sql/sem/catid.DescID"];
>   uint32 index_id = 2 [(gogoproto.customname) = "IndexID", (gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/sql/sem/catid.IndexID"];
>   uint32 zone_config_id = 3 [(gogoproto.customname) = "ZoneConfigID"];
> ```

Is there such a thing as a zone config ID? I thought they were keyed by the descriptor ID in the `system.zones` table.
___
*[`pkg/sql/schemachanger/scplan/internal/opgen/opgen_table_subzone_config.go` line 27 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/83535#-N5q5Bc9B27UrdgHuvLH:-N5q5Bc9B27UrdgHuvLI:b-vjffbm) ([raw file](https://github.com/cockroachdb/cockroach/blob/3765f1bb1802a485f8d0f562900d05d1f5d9dc19/pkg/sql/schemachanger/scplan/internal/opgen/opgen_table_subzone_config.go#L27)):*
> ```Go
> 						DescriptorID:  this.TableID,
> 						IndexID:       this.IndexID,
> 						ZoneConfig:    this.ZoneConfig,
> ```

Deep-copy this protobuf and the one below with `protoutil.Clone` just to stay on the safe side.
___
*[`pkg/sql/schemachanger/screl/attr.go` line 310 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/83535#-N5q3ByF5M2HlxQUwf-3:-N5q3ByF5M2HlxQUwf-4:b-x1icmd) ([raw file](https://github.com/cockroachdb/cockroach/blob/3765f1bb1802a485f8d0f562900d05d1f5d9dc19/pkg/sql/schemachanger/screl/attr.go#L310)):*
> ```Go
> 		rel.EntityAttr(IndexID, "IndexID"),
> 		rel.EntityAttr(ZoneConfigID, "ZoneConfigID"),
> 		rel.EntityAttr(PartitionName, "PartitionName"),
> ```

We should be careful not to define too many `Attr` values as our supply is quite constrained. Perhaps we could re-use `Name` here instead of having to introduce a `PartitionName`? I'm not sure.
___
*[`pkg/sql/schemachanger/scplan/testdata/drop_type` line 29 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83535#-N5psqEJ8h44GO6H_wzn:-N5psqEJ8h44GO6H_wzo:b-if047o) ([raw file](https://github.com/cockroachdb/cockroach/blob/002b0d9a410941931a4ad87488f645d53321e105/pkg/sql/schemachanger/scplan/testdata/drop_type#L29)):*
> ```
>       DescID: 104
>     *scop.NotImplemented
>       ElementType: scpb.EnumTypeValue
> ```

This shouldn't be here, you're missing an op-rule to tell the planner to skip removing a single enum value when dropping the whole type.

I'm surprised this isn't causing more test failures, to be honest. I filed https://github.com/cockroachdb/cockroach/issues/83672 to help in this regard.


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83535)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @fqazi, and @postamar)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/region_utils.go` line 63 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/83535#-N5q8sWh9UTQbSZrK0rE:-N5qAz8Q0_Go1tQ0P1lp:b-y0mvnl) ([raw file](https://github.com/cockroachdb/cockroach/blob/3765f1bb1802a485f8d0f562900d05d1f5d9dc19/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/region_utils.go#L63)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I don't know what's the right thing to do but I'm pretty sure building whole zone config protos in `scbuildstmt` isn't it. Sorry if this comment isn't practically helpful in the immediate. I think this is a sign that zone config support needs to be better thought out first.
</blockquote></details>

This should have been pushed down into the execution phase when adding a new subzone (inside the mutation state visitor). Think this was an oversight on my part, I only moved the span-related parts over.
___
*[`pkg/sql/schemachanger/scdecomp/decomp.go` line 301 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/83535#-N5q7Q1T6j6-QwMoJKf9:-N5qBiCP6CUr_SfCZwVw:b-59wzrc) ([raw file](https://github.com/cockroachdb/cockroach/blob/3765f1bb1802a485f8d0f562900d05d1f5d9dc19/pkg/sql/schemachanger/scdecomp/decomp.go#L301)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Since these elements now show up for any table which have zone configs, how do we know that they're correctly handled when we `DROP TABLE`? In practice they end up getting removed by the GC job and that's fine, but you should have op-rules to ensure that these elements don't emit any ops in such a scenario. This is currently missing.

I think you should break down this commit further, and perhaps this PR further as well, to add zone config support in a first step (and check that you pass all the tests) and in a subsequent step worry about adding columns in multi-region tables.
</blockquote></details>

Shouldn't the generic rule inside dep_drop.go also trim these out? We shouldn't need an explicit rule to filter them out?

In terms of zone config support, I think we should have a narrower protobuf here for initial support for add column. Doing a decomposition here is not trivial and not worth the effort until we bring up more things (i.e. nothing will use the elements and it will be noise in the graph assuming we get the model right). For the purpose of ADD COLUMNS / INDEXES we only care about subzones, which are already split out.
___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 490 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/83535#-N5q7ATuDtzNYjOotQ47:-N5qDqf41ZFCwjleBeBR:b9w6i46) ([raw file](https://github.com/cockroachdb/cockroach/blob/3765f1bb1802a485f8d0f562900d05d1f5d9dc19/pkg/sql/schemachanger/scpb/elements.proto#L490)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Is there such a thing as a zone config ID? I thought they were keyed by the descriptor ID in the `system.zones` table.
</blockquote></details>

There are intentionally allocated inside the builder to help support transactions, where elements can be added or replaced based on the IDs involved.


<!-- Sent from Reviewable.io -->

chengxiong-ruan commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_column.go:
```diff
@@ -232,8 +232,43 @@ func alterTableAddColumn(
+		tableElts := b.QueryByID(tbl.TableID)
+		_, _, tblNS := scpb.FindNamespace(tableElts)
+		if _, _, elem := scpb.FindTableLocalityRegionalByRow(tableElts); elem != nil {
+			_, _, currentZoneConfigElem := scpb.FindTableZoneConfig(tableElts)
```
I wasn't able to find where `scpb.FindTableZoneConfig` is defined either in master or in this pr :(
But...can a multiregion table have zero or more than one zone config?

chengxiong-ruan replied:
nvm...my bad..github collapses the file heh..

chengxiong-ruan commented on pkg/sql/schemachanger/scbuild/builder_state.go:
```diff
@@ -173,6 +175,18 @@ func (b *builderState) CurrentUserHasAdminOrIsMemberOf(role username.SQLUsername
+func (b *builderState) NextZoneConfigID(table *scpb.Table) uint32 {
+	tblElts := b.QueryByID(table.TableID)
+	zoneConfigID := uint32(0)
+	scpb.ForEachTableZoneConfig(tblElts, func(current scpb.Status, target scpb.TargetStatus, e *scpb.TableZoneConfig) {
```
sorry... I couldn't find where `ForEachTableZoneConfig` function is defined either in master and in this pr... Would a multi-region table have more than one zone config? Or `ForEachTableZoneConfig` function is actually also looping through all the sub-zone configs as well?

fqazi committed:
sql/schemachanger: adopt metadata updater for zone config updates

Previously, the code to update the zone configs was separate
between the declarative schema changer and the legacy schema
changer. This was inadequate because we have a opportunity
to share logic here. To address this the logic to update and
delete zone configurations is now shared.

Release note: None

fqazi committed:
sql/schemachanger: adopt zone config reader

Previously, the zone config reader code was only used
by the declarative schema changer. But, we can take
advantage of logic in existing zone config logic. To address
this, this patch will adopt the zone config reader in existing
logic.

Release note: None

fqazi committed:
sql/regionconfig: move SurvivalGoal and DataPlacement to catpb

Previously, SurvivalGoal and DataPlacement were located inside,
the descpb protobuf package. To help support using these enums
inside the declarative schemachanger we will move these enums
into the catpb package.

Release note: None

fqazi committed:
sql/schemachanger: partially decompose zone config elements

Previously, our element model did not decompose zone config
and multi-region related elements for the declarative schema
changer. These changes add support for partially decomposing these
elements, so that we can issue proper subzone related updates
when creating indexes during add column.

Release note: None

fqazi committed:
sql/schemachanger: support add column on multiregion tables

Previously, the declarative schema changer would fall back
to the legacy schema changer when multiregion tables were
encountered. This was inadequate because this meant that
add column wasn't well supported in multiregion databases.
To address this patch will add support for updating zone
configs as a part of the ADD COLUMN.

Release note: None

postamar commented:
At this point this PR has a sizeable fraction of the whole engineering org tagged in it despite not being ready for review any time soon. I strongly encourage you to:
1. close this PR;
2. open any smaller spin-off PR instead for the more uncontroversial changes and get them merged quickly independently of the rest: enum value element, metadataUpdater refactors, etc.
3. open a multi-commit draft PR with:
  - the foundational and less obvious changes which are much more likely to be subject to revision, namely the zone config support in the element model, scdecomp and rules, and making sure that some of the DROP test cases involve zone configs but don't actually do anything with them;
  - perhaps some changes which add some limited CONFIGURE ZONE support & accompanying tests;
  - multi-region ADD COLUMN support & accompanying tests.

fqazi commented:
@postamar Pretty much that's the plan the individual bits are independent, and I'll cherry pick and open PR's. The bloat here is mainly from the refactors for catpb and regionutils (those should be uncontroversial).

postamar commented:
Thanks. Yes, I'll expect those will sail through, which they should: I'm worried about these growing stale over time.

I feel that the scope of the changes required here is not clear at this time. I'm looking at `set_zone_config.go` in particular and wondering how much of that we should add support for in order to not foot-gun ourselves when designing the element model for the zone configs. This chicken-and-egg problem has been a recurrent challenge with all recent declarative schema changer feature work, which ended up requiring many iterations, so I'm growing a bit paranoid about scope. This isn't your fault, it just so happens that every time anyone pulls on a thread it ends up being a large effort.
